![](doc/images/CAW_logo.png)
# Cancer Analysis Workflow Results Delivery 
This README describes the delivery directory structure for files passed to users at [NGI][ngi-link]

There are four sections dedicated for different results: [Annotation], Preprocessing, Reports and 
VariantCalling. All the four sections can have sub-directories containing results from different software.

## Annotation: 

This directory contains results from the final annotation steps: two software are used for annotation, [VEP][vep-link] and [snpEff][snpeff-link]. 
Only a subset of the VCF files are annotated, and only variants that have a PASS filter. FreeBayes results are not annotated in the moment yet as
we are lacking a decent somatic filter. For HaplotypeCaller the germline variations are annotated for both the tumour and the normal sample.

All the VCFs annotated have an `ann.vcf` extension,  and a summary HTML file associated. 


## Preprocessing:

### NonRealigned:

This directory is usually empty, as it is a placeholder for the original mapped, merged and duplicate marked BAM files. After these steps the BAM files are 
processed further, reads are realigned around known indels, and recalibrated.

### NonRecalibrated:

This is the place for the BAM file delivered to users: besides the realigned files the recalibration tables are also stored (`*.recal.table`), these can be
used to create base recalibrated files. The `.tsv` file is autogenerated also, these can be used by CAW for further processing and/or variant calling. 

The BAM file headers contain the 

### Recalibrated:

This directory is usually empty, it is the location for the final recalibrated files in the preprocessing pipeline: recalibrated BAMs are usually 2-3 times 
larger than the realigned files, and are needed only by MuTect1 and MuTect2 (considering GATK 3.8). To re-generate recalibrated BAMs you have to apply the 
recalibration table delivered to the `NonRecalibrated` directory either by calling CAW, or doing this [recalibration step][BQSR-link] yourself.

## Reports:
### bamQC  
### BCFToolsStats  
### FastQC  
### MarkDuplicates  
### MultiQC  
### SamToolsStats


## VariantCallings:

[BQSR-link]: https://gatkforums.broadinstitute.org/gatk/discussion/44/base-quality-score-recalibration-bqsr
[ngi-link]: https://ngisweden.scilifelab.se/
[snpeff-link]: http://snpeff.sourceforge.net/
[vep-link]: http://www.ensembl.org/Tools/VEP
[logo]: ttps://img.shields.io/github/release/SciLifeLab/CAW.svg

